sql/atxs/atxs.go

```diff
@@ -801,23 +811,39 @@ func CountAtxsByOps(db sql.Executor, operations builder.Operations) (count uint3
 func IterateForGrading(
 	db sql.Executor,
 	epoch types.EpochID,
-	fn func(id types.ATXID, atxtime, prooftime int64, weight uint64) bool,
+	fn func(id types.ATXID, atxTime, proofTime int64, weight uint64) bool,
 ) error {
 	if _, err := db.Exec(`
-		select atxs.id, atxs.received, identities.received, effective_num_units, tick_count from atxs
-		left join identities on atxs.pubkey = identities.pubkey
-		where atxs.epoch == ?1;`,
-		func(stmt *sql.Statement) {
-			stmt.BindInt64(1, int64(epoch))
-		}, func(stmt *sql.Statement) bool {
-			id := types.ATXID{}
-			stmt.ColumnBytes(0, id[:])
-			atxtime := stmt.ColumnInt64(1)
-			prooftime := stmt.ColumnInt64(2)
-			units := uint64(stmt.ColumnInt64(3))
-			ticks := uint64(stmt.ColumnInt64(4))
-			return fn(id, atxtime, prooftime, units*ticks)
-		}); err != nil {
+		SELECT atxs.id, atxs.received, effective_num_units, tick_count, identities.received, malfeasance.received
+		FROM (
+			SELECT * FROM atxs WHERE epoch = ?1
+		) AS atxs
+		LEFT JOIN identities ON atxs.pubkey = identities.pubkey
+		LEFT JOIN malfeasance ON atxs.pubkey = malfeasance.pubkey
```

Comment:
Why change to this form of `SELECT ... FROM (SELECT * FROM atxs ...)`? Is it more efficient?

Reply:
I just updated the PR description to explain 😅. Instead of joining the `identities` and `malfeasance` table with the ATXs table and then selecting based on epoch this query now filters all ATXs by the epoch we are interested in and then joins the result before selecting the columns we care about. This should be much faster because we are joining fewer entries.

Reply:
Alright, it makes sense :+1:. One more question: why make a join here, while in other places you prefer a separate query to check malfeasance?

Reply:
Good that you asked again! Actually both (the nested query and the non nested one) should be about the same speed because modern SQL should know that applying `WHERE` before `JOIN` here is much faster. I tested it again on the `state.sql` of a mainnet node (before only a testnet node where there wasn't much difference) and the non-nested query is actually faster; probably because the nested one has to materialize the temporary result which is costly. I'll remove the nested queries again.

Regarding separate check: the problem is that these are streaming/iterating calls. The caller would need to query for malfeasance in the callback of this function which I believe can lead to deadlocks. I'll see if I can make it work somehow.

Reply:
`atxs.IterateAtxIDsWithMalfeasance` is only used for v1 GRPC API. `atxs.IterateForGrading` and `atxs.IterateAtxsWithMalfeasance` are used for activeset and beacon generation respectively. Both are currently not called because of the bootstrapped beacon and active set.

Also I am not sure if splitting these queries into 1 query for ATXs and multiple for malfeasance would speed them up. What probably needs to happen for both - activeset and beacon generation - is that these should be prepared eagerly when new ATXs come in and not lazily the first time they are needed. This would slow down node starting times (more work up front) but would reduce load on the DB from those two components by quite a bit.

Reply:
Maybe but for activeset and beacon we don't only need to know if someone is malicious but also since when we consider them so, so we would need to update `atxsdata` accordingly.